src/consensus.rs

```diff
@@ -822,92 +760,228 @@ fn handle_committed_entries(
     Ok(stop_consensus)
 }
 
-async fn who_is(
-    peer_id: collection::shards::shard::PeerId,
+struct RaftMessageBroker {
+    senders: HashMap<PeerId, Sender<RaftMessage>>,
+    runtime: Handle,
     bootstrap_uri: Option<Uri>,
-    config: Arc<ConsensusConfig>,
     tls_config: Option<ClientTlsConfig>,
-) -> anyhow::Result<Uri> {
-    log::debug!("Resolving who is {peer_id}");
-    let bootstrap_uri =
-        bootstrap_uri.ok_or_else(|| anyhow::anyhow!("No bootstrap uri supplied"))?;
-    let bootstrap_timeout = Duration::from_secs(config.bootstrap_timeout_sec);
-    // Use dedicated transport channel for who_is because of specific timeout
-    let channel = make_grpc_channel(
-        bootstrap_timeout,
-        bootstrap_timeout,
-        bootstrap_uri,
-        tls_config,
-    )
-    .await
-    .map_err(|err| anyhow!("Failed to create timeout channel: {}", err))?;
-    let mut client = RaftClient::new(channel);
-    Ok(client
-        .who_is(tonic::Request::new(GrpcPeerId { id: peer_id }))
-        .await?
-        .into_inner()
-        .uri
-        .parse()?)
+    consensus_config: Arc<ConsensusConfig>,
+    consensus_state: ConsensusStateRef,
+    transport_channel_pool: Arc<TransportChannelPool>,
+}
+
+impl RaftMessageBroker {
+    pub fn new(
+        runtime: Handle,
+        bootstrap_uri: Option<Uri>,
+        tls_config: Option<ClientTlsConfig>,
+        consensus_config: ConsensusConfig,
+        consensus_state: ConsensusStateRef,
+        transport_channel_pool: Arc<TransportChannelPool>,
+    ) -> Self {
+        Self {
+            senders: HashMap::new(),
+            runtime,
+            bootstrap_uri,
+            tls_config,
+            consensus_config: consensus_config.into(),
+            consensus_state,
+            transport_channel_pool,
+        }
+    }
+
+    pub fn send(&mut self, messages: impl IntoIterator<Item = RaftMessage>) {
+        for message in messages {
+            let peer_id = message.to;
+
+            let sender = match self.senders.get(&peer_id) {
+                Some(sender) => sender,
+                None => {
+                    log::debug!("Spawning message sender task for peer {peer_id}...");
+
+                    let (task, handle) = self.message_sender();
+                    let future = self.runtime.spawn(task.exec());
+                    drop(future); // drop `JoinFuture` explicitly to make clippy happy
+
+                    self.senders.insert(peer_id, handle);
+
+                    self.senders
+                        .get(&peer_id)
+                        .expect("message sender task spawned")
+                }
+            };
```

Comment:
Though this is fine, we can replace this pattern with [`get_or_insert`](https://doc.rust-lang.org/std/option/enum.Option.html#method.get_or_insert_with)

Reply:
Not really. `get_or_insert` is the method on `Option`. The equivalent here is [`entry`] API on `HashMap`... but, tbh, I find explicit match with `None` branch easier to read than `hash_map.entry(key).or_insert_with(|| produce_value())`. :/

[`entry`]: https://doc.rust-lang.org/std/collections/hash_map/struct.HashMap.html#method.entry

Reply:
I agree. Thanks for looking into it.